src/Core/ExecutionPathTracer/ExecutionPathTracer.cs

```diff
@@ -0,0 +1,307 @@
+            typeof(Microsoft.Quantum.Canon.ApplyToEachC<Qubit>),
+            typeof(Microsoft.Quantum.Canon.ApplyToEachA<Qubit>),
+            typeof(Microsoft.Quantum.Canon.ApplyToEachCA<Qubit>),
+        };
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ExecutionPathTracer"/> class.
+        /// </summary>
+        /// <param name="depth">
+        /// The depth at which to render operations.
+        /// </param>
+        public ExecutionPathTracer(int depth = 1) => this.renderDepth = depth + 1;
+
+        /// <summary>
+        /// Returns the generated <c>ExecutionPath</c>.
+        /// </summary>
+        public ExecutionPath GetExecutionPath()
+        {
+            var qubits = this.qubitRegisters.Keys
+                .OrderBy(k => k)
+                .Select(k =>
+                {
+                    var qubitDecl = new QubitDeclaration(k);
+                    if (this.classicalRegisters.ContainsKey(k))
+                    {
+                        qubitDecl.NumChildren = this.classicalRegisters[k].Count;
+                    }
+
+                    return qubitDecl;
+                })
+                .ToArray();
+
+            return new ExecutionPath(qubits, this.operations.ToArray());
+        }
+
+        /// <summary>
+        /// Provides the event listener to listen to <c>SimulatorBase</c>'s <c>OnOperationStart</c> event.
+        /// </summary>
+        public void OnOperationStartHandler(ICallable operation, IApplyData arguments)
+        {
+            // If the operation type is one of the nestedTypes, go one depth deeper and parse
+            // those operations instead
+            if (this.nestedTypes.Contains(operation.GetType())) return;
+
+            this.currDepth++;
+
+            // Parse operations at specified depth
+            if (this.currDepth == this.renderDepth)
+            {
+                var parsedOp = this.ParseOperation(operation, arguments);
+                if (parsedOp != null)
+                {
+                    this.operations.Add(parsedOp);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Provides the event listener to listen to <c>SimulatorBase</c>'s <c>OnOperationEnd</c> event.
+        /// </summary>
+        public void OnOperationEndHandler(ICallable operation, IApplyData result)
+        {
+            if (this.nestedTypes.Contains(operation.GetType())) return;
+            this.currDepth--;
+        }
+
+        private static bool IsPartialApplication(ICallable operation)
+        {
+            var t = operation.GetType();
+            if (t == null) return false;
+
+            return t.IsGenericType && t.GetGenericTypeDefinition() == typeof(OperationPartial<,,>);
+        }
+
+        /// <summary>
+        /// Retrieves the <c>QubitRegister</c> associated with the given <c>Qubit</c> or create a new
+        /// one if it doesn't exist.
```

Comment:
It's a bit confusing for a "Get★" method to have a side-effect of creating something...

Reply:
Would `GetOrCreate*` be better?

Reply:
In general I agree with @cgranade's comment, but in this particular case I like `Get*`, since it's basically just allowing a lazy-initialization of the `qubitRegisters` dictionary. Up to you, though, no big deal either way.

Reply:
Cool, thanks for the input! I'll stick to `Get*` unless there are strong objections against this. :)